core/Enclave/enclave_hsm.cpp

```diff
@@ -294,43 +327,92 @@ sgx_status_t enclave_create_rsa_key(uint8_t *cmk_blob,
         printf("ecall create_rsa_key cmk_keyblob_size:%lu < key_blob_size:%d.\n", cmk_blob_size, real_keyblob_size);
         return SGX_ERROR_INVALID_PARAMETER;
     }
-        rsa_key_data_t *keyblob = (rsa_key_data_t *)cmk_blob;
+    if (keyspec == NULL || padding_mode == NULL)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    RSA     *rsa            = NULL;
+    BIGNUM  *bne            = NULL;
+    BIO     *public_key     = NULL;
+    BIO     *private_key    = NULL;
+    char    *p_public_key   = NULL;
+    char    *p_private_key  = NULL;
+    string payload;
      
     do {
-        RSA *rsa = RSA_new();
-        BIGNUM *bne = BN_new();
-        BN_set_word(bne, RSA_F4);
-        switch (keyspec)
-        {
+        // init rsa key struct
+        rsa = RSA_new();
+        if (rsa == NULL) {
+            ret = SGX_ERROR_OUT_OF_MEMORY;
+            break;
+        }
+
+        bne = BN_new();
+        if (bne == NULL) {
+            ret = SGX_ERROR_OUT_OF_MEMORY;
+            break;
+        }
+
+        if(!BN_set_word(bne, RSA_F4)) {
+            break;
+        }
+
+        // generate rsa key
+        int res;
+        switch (keyspec) {
             case EH_RSA_2048:
-                RSA_generate_key_ex(rsa, RSA_2048_KEY_BITS, bne, nullptr);
+                res = RSA_generate_key_ex(rsa, RSA_2048_KEY_BITS, bne, nullptr);
                 break;
             case EH_RSA_3072:
-                RSA_generate_key_ex(rsa, RSA_3072_KEY_BITS, bne, nullptr);
+                res = RSA_generate_key_ex(rsa, RSA_3072_KEY_BITS, bne, nullptr);
+                break;
+            case EH_RSA_4096:
+                res = RSA_generate_key_ex(rsa, RSA_4096_KEY_BITS, bne, nullptr);
+                break;
+            default:
                 break;
         }
-        
-        BIO *public_key = BIO_new(BIO_s_mem());
-        BIO *private_key = BIO_new(BIO_s_mem());
+        if(!res) {
+            break;
+        }
 
-        PEM_write_bio_RSAPrivateKey(private_key, rsa, EVP_des_ede3_ofb(), (unsigned char*)g_domain_key, sizeof(g_domain_key), NULL, NULL);
+        // storage rsa key with pem format
+        public_key = BIO_new(BIO_s_mem());
```

Comment:
why still malloc it in the enclave?

as i mentioned that you need to do it by the following steps:
1. calculate the memory size for the rsa key size
2. malloc it in the provider layer, and then pass the memory buffer into the enclave
3. enclave just check the memory is valid, and use it to store the new generated rsa keys.

Reply:
Here is just use some memory to storage public key and private key and use as the buffer for keyblob with aes_gcn_128 encryption 
the space will be free in enclave